Support quantization in Dynamo, ONNX and ET exporters (#47747)
huggingface/transformers: docs/source/en/exporters.md

</details>

## Quantization

Every export config accepts a `quantizer`. Set it to any PT2E

> [`Quantizer`](https://docs.pytorch.org/ao/stable/) and the exporter runs post-training quantization on

Reviewer: ```suggestion
[`Quantizer`](https://docs.pytorch.org/ao/main/pt2e_quantization/index.html) and the exporter runs post-training quantization on
```

the traced graph (`prepare_pt2e` → calibrate → `convert_pt2e`) before the program is returned or
lowered. Quantization happens on the graph rather than the modeling code, so a single `quantizer` works

> across every backend and architecture without per-model handling.

Reviewer: hmm the code example below seems a bit contradictory with the `NOTE` at the end? should we maybe use `export_for_generation` instead in the example or lead with the constraint?

```suggestion
across every backend and architecture without per-model handling.

Quantize through [`~HfExporter.export_for_generation`], which exports the decomposed generation components. Their attention mask is a precomputed graph input, which keeps PT2E away from the in-graph mask construction that trips its `make_fx` retrace on a full model forward.
```

```python
from transformers import LlamaForCausalLM
from transformers.exporters import DynamoExporter, DynamoConfig
from torchao.quantization.pt2e.quantizer.x86_inductor_quantizer import (
    X86InductorQuantizer,
    get_default_x86_inductor_quantization_config,
)

model = LlamaForCausalLM.from_pretrained("meta-llama/Llama-3.2-1B").eval()
inputs = ...  # forward kwargs

quantizer = X86InductorQuantizer().set_global(get_default_x86_inductor_quantization_config())
config = DynamoConfig(dynamic=True, quantizer=quantizer, calibration_dataset=[inputs])
exported = DynamoExporter().export(model, inputs, config)  # quantize/dequantize ops folded into the graph
```

### Choosing a quantizer

Each target runtime expects its own quantizer. The quantized graph then runs on inductor (int8),
translates to ONNX `QuantizeLinear`/`DequantizeLinear` (per-channel included), or lowers to an

> ExecuTorch `.pte`.

Reviewer: ```suggestion
Each target runtime expects its own quantizer. Whichever you pass, the quantized graph is portable from there. It runs on inductor as int8, translates to ONNX `QuantizeLinear`/`DequantizeLinear` (per-channel included), or lowers to an ExecuTorch `.pte`.
```

> | Where you'll run | Quantizer to pass |

Reviewer: would it be helpful to add a third column that tells you where to import it from?

| --- | --- |
| PyTorch inductor, or ONNX Runtime (QDQ) | `X86InductorQuantizer` (torchao) |
| ExecuTorch XNNPACK backend | `XNNPACKQuantizer` |

> | ExecuTorch QNN backend (Qualcomm HTP) | `QnnQuantizer` |

Reviewer 2: Although the biggest use case here is HTP, we do also have quantizer for other hardware on the SoC, such as LPAI. Maybe making this more general?

Author: not really knowledgeable on this matter so please don't hesitate to suggest changes. i asked claude to make it more general, tell if that works for you.

Reviewer 2: LGTM. Mentioning this for mainly awareness 😊

### Calibration

> `calibration_dataset` is a list of forward-kwarg dicts run through the prepared graph to gather

Reviewer 2: Have we considered using something like a DataLoader here? I'm asking because I'd like to better understand how the framework is intended to handle different types of datasets, both now and in the future.

Author: changed it to an iterable which should work with dataloaders as well

Author: this is a very minimal integration, more features can be added later 🤗

Reviewer 2: Makes sense. Thanks for the explanation!

observer statistics. Omit it and the exporter falls back to a single pass over the export's own sample
inputs, with a warning (one sample can skew the observed ranges).

For generative models, give the dataset to [`~HfExporter.export_for_generation`] instead of the
exporter directly. You pass one generate-style dataset; it runs through a short `generate` to capture
each component's real inputs, and `prefill` / `decode` / the encoders are each calibrated on their own

> captured activations.

Reviewer: this kind of reads like pass the `calibration_dataset` to `export_for_generation`. may be clearer like:

```suggestion
For generative models, set `calibration_dataset` on the config you pass to [`~HfExporter.export_for_generation`] and give it generate-style kwargs. Each sample runs through a short `generate`, and every component (`prefill`, `decode`, the encoders) is calibrated on the activations captured for it.
```
